HDDS-8035. Intermittent timeout in TestOzoneManagerHAWithData.testOMHAMetrics

xBis7 opened the pull request:
## What changes were proposed in this pull request?

In #4140 , we added an integration test, `testOMHAMetrics` where we restart the leader OM and then wait for a new leader to be elected in order to check the metrics during leader change. A timeout, set to 80000 millis, was added on waiting for the new leader. For the OM failover, the failover can last minutes and therefore we need to expand the timeout for the test.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-8035

## How was this patch tested?

This patch was tested manually locally. `@Test` annotation was replaced with `@RepeatedTest(100)` and the test was run with Maven locally. With the original timeout set to 80000 millis the test failed, while with the increased timeout set to 300000 millis the test passed all 100 repetitions.

```
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running org.apache.hadoop.ozone.om.TestOzoneManagerHAWithData
[INFO] Tests run: 107, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 444.62 s - in org.apache.hadoop.ozone.om.TestOzoneManagerHAWithData
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 107, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
```

xBis7 commented:
@neils-dev @adoroszlai Can you please take a look at this PR?

xBis7 commented:
> Are there any timeouts related to OM Ratis that could be tweaked to reduce the time needed for leader election after restart?

@adoroszlai I've been looking into `OzoneManagerRatisServer.newRaftProperties()` and `OMConfigKeys` and I can't see something that could make a difference. I tweaked some of them and then retried the tests but there was no change. Do you have something specific in mind?

adoroszlai reviewed (changes requested):
@xBis7 I have run 100x iterations of the test and it has timed out in 30% of repetitions even with 300s limit.

https://github.com/adoroszlai/hadoop-ozone/actions/runs/4393202042

xBis7 commented:
@adoroszlai Thanks for testing it. I'll track it down and fix it.

neils-dev commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOzoneManagerHAWithData.java:
```diff
@@ -107,7 +107,6 @@ public void testMultipartUpload() throws Exception {
-  @Flaky("HDDS-8035")
   public void testOMHAMetrics() throws InterruptedException,
       TimeoutException, IOException {
     waitForLeaderToBeReady();
```
Should also fix `waitForLeaderToBeReady()`.  Currently unnecessarily checking `isLeaderReady() `within the method.  
Specifically, the line 
https://github.com/apache/ozone/blob/ce418581a8b4533a47de741d543072903562a4c6/hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOzoneManagerHAWithData.java#L181
`isLeaderReady() `is redundant as it is called within the `getCluster().getOMLeader()`
https://github.com/apache/ozone/blob/ce418581a8b4533a47de741d543072903562a4c6/hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/MiniOzoneHAClusterImpl.java#L197

xBis7 replied:
@neils-dev I agree and apart from that, I've also verified that during OM restart, there is no leader change taking place.

Instead of restarting the leader, we should call `getCluster().restartOzoneManager();` which restarts all OMs and triggers a new leader election.

xBis7 commented:
@adoroszlai Although, the test always passes when I run it locally on repeat, when I use the workflow you shared, it fails more often than it succeeds. 

I have also confirmed that during repetitions there is no leader change. On the first repetition the leader is always Node-1 and it changes to Node-3 and for every other repetition it's always Node-3.

I've added a check that all three OMs are up and running. I've set maximum timeout to the amount of time we wait for the Ratis failover. I don't think that the timeout should be more than that. Furthermore, I've changed the check

```diff
- getCluster().getOMLeader().isLeaderReady();
+ getCluster().getOMLeader(); 
```
as `getOMLeader()` also checks if the leader is ready and returns null if not. Check [here](https://github.com/apache/ozone/blob/master/hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/MiniOzoneHAClusterImpl.java#L194-L208).


There is an underlying issue here that causes the timeout and it might even have to do with the MiniCluster and its workings. I don't have the time to investigate more and we have other priorities. I plan to convert this into a draft PR and maybe close it later on.

adoroszlai commented:
Thanks @xBis7 for the update.

> when I use the workflow you shared, it fails more often than it succeeds.

That may happen because Github runners have less resources than your dev environment, or because of the difference in repetitions (repeating a single test method vs. the entire test class).

> I don't have the time to investigate more and we have other priorities.

No problem, we can keep it marked as `@Flaky`.

xBis7 commented:
@adoroszlai I can put back the `@Flaky` annotation and we can keep the improvements for whoever picks up fixing it.

adoroszlai commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOzoneManagerHAWithData.java:
```diff
@@ -109,8 +109,10 @@ public void testMultipartUpload() throws Exception {
   @Test
-  public void testOMHAMetrics() throws InterruptedException,
-      TimeoutException, IOException {
+  public void testOMHAMetrics() throws Exception {
+    shutdown();
+    init();
+    getCluster().waitForClusterToBeReady();
```
Instead of such tricks, I think it's better to move this test to a separate class.

xBis7 replied:
You are right. I made the changes.

adoroszlai reviewed (approved):
Thanks @xBis7 for persisting with this issue.

100/100 passed:
https://github.com/adoroszlai/hadoop-ozone/actions/runs/4617974982

xBis7 commented:
@adoroszlai Thanks for reviewing this patch!
